patroni/config_generator.py

```diff
@@ -0,0 +1,289 @@
+
+from .config import Config
+from .exceptions import PatroniException
+from .postgresql.config import ConfigHandler
+from .postgresql.misc import postgres_major_version_to_int
+from .utils import get_major_version, patch_config
+
+
+def get_ip() -> str:
+    """
+    Try to get an ip address of the hostname returned by :func:`gethostname`.
+
+    .. note::
+            Can also return local ip
+
+    :returns: the first element in the sorted list of the addresses returned by :func:`getaddrinfo`.
+              Sorting guarantees it will prefer ipv4.
+    """
+    hostname = None
+    try:
+        hostname = socket.gethostname()
+        return sorted(socket.getaddrinfo(hostname, 0, socket.AF_UNSPEC, socket.SOCK_STREAM, 0),
+                      key=lambda x: x[0])[0][4][0]
+    except OSError as e:
+        sys.exit(f'Failed to define ip address: {e}')
+    except IndexError:
+        sys.exit(f'Failed to define ip address. No address returned by getaddrinfo for {hostname}')
+
+
+def get_int_major_version(config: Optional[Dict[str, Any]] = None) -> int:
+    """Get major version of PostgreSQL from the binary as an integer.
+
+    :param config: optional dictionary representing Patroni config. If contains bin_dir and/or the custom
+    bin_name for postgres, the values are used for the version retrieval.
+    """
+    config = config or {}
+    postgres_bin = ((config.get('postgresql') or {}).get('bin_name') or {}).get('postgres', 'postgres')
+    return postgres_major_version_to_int(get_major_version(config['postgresql'].get('bin_dir') or None, postgres_bin))
+
+
+def get_bin_dir_from_running_instance(data_dir: str) -> str:
+    """Define the directory postgres binaries reside using postmaster's pid executable."""
+    postmaster_pid = None
+    try:
+        with open(f"{data_dir}/postmaster.pid", 'r') as f:
+            postmaster_pid = f.readline()
+            if not postmaster_pid:
+                sys.exit('Failed to obtain postmaster pid from postmaster.pid file')
+            postmaster_pid = int(postmaster_pid.strip())
+    except OSError as e:
+        sys.exit(f'Error while reading postmaster.pid file: {e}')
+    import psutil
+    try:
+        return os.path.dirname(psutil.Process(postmaster_pid).exe())
+    except psutil.NoSuchProcess:
+        sys.exit('Obtained postmaster pid doesn\'t exist')
+
+
+def enrich_config_from_running_instance(config: Dict[str, Any], no_value_msg: str, dsn: Optional[str] = None) -> None:
+    """Extend the passed ``config`` dictionary with the values gathered from a running instance.
+
+    Get
+    - non-internal GUC values having configuration file, postmaster command line or environment variable as a source
+    - postgresql.connect_address, postgresql.listen,
+    - postgresql.pg_hba and postgresql.pg_ident if hba_file/ident_file is set to the default value
+    - superuser auth parameters (from the options used for connection)
+    And redefine scope with the clister_name GUC value if set
```

Comment:
```suggestion
    - superuser auth parameters (from the options used for connection)

    And redefine scope with the ``cluster_name`` GUC value if set.
```

We also need a space after the list.